### PR TITLE
Prevent Overwriting of User's Attributes on Register (#897)

### PR DESCRIPTION
**Pull request by dav-is**
I read through the entire SSO process and this is the only security issue I could even remotely find regarding to flarum/issue-archive#332. This will prevent any of the user's attributes to be overwritten. For example, if a SSO provider neglected to verify emails.

### CODE REVIEW

File: src/Core/Command/RegisterUserHandler.php [hunk 1]
```diff
@@ -115,7 +116,9 @@ public function handle(RegisterUser $command)
         // from the get-go.
         if (isset($token)) {
             foreach ($token->payload as $k => $v) {
-                $user->$k = $v;
+                if ($user->$k == '' || ! isset($user->$k)) {
```
Comment (franzliedke) [inline src/Core/Command/RegisterUserHandler.php]: Isn't this the same as `if (empty($user->$k)`?

Comment (dav-is) [inline src/Core/Command/RegisterUserHandler.php]: Well empty would return true for 0 or false

Comment (franzliedke) [inline src/Core/Command/RegisterUserHandler.php]: So does `== ''`, see https://3v4l.org/Z6cl0.

Please use `===` then. :)

**Comment by tobscure** [comment]
Can you please give a more concrete example of exactly how this could be exploited to screw something up? I agree that there's probably an issue here, but I'm having trouble imagining what that might be, and the complexity of the `if ($user->$k === '' || ! isset($user->$k))` condition is setting off alarm bells for me.

**Comment by dav-is** [comment]
A user's attribute for email_when _mentioned (just making this up) is set to `0`. This setting is set and has a value of false. Therefore, it should not be tampered with. `if (empty($user->$k))` would return true because the value is false, but we still don't want to allow our SSO to modify the set variable.

**Review by luceos — CHANGES_REQUESTED** [review]
the conditional can be shortened using an empty() function

**Comment by dav-is** [comment]
@luceos Did you read my rational behind not using empty()?

**Comment by luceos** [comment]
I skipped your last comment, my apologies. Even more impressed 😄

**Review by luceos — APPROVED** [review]

**Comment by Luceos** [comment]
@tobscure or @franzliedke either merge this or assign to a milestone please. After having done that please unassign both of you.

**Comment by franzliedke** [comment]
Thanks for the PR! And your patience. ;)

**Comment by franzliedke** [comment]
I [slightly tweaked the condition](https://github.com/flarum/core/commit/7c1b0bfcf2fccc361cc984375dcc6a6430216b3a) to a statement that signals the intent slightly better - at least in my opinion.

And damn, should have squashed this ;)

**Comment by franzliedke** [comment]
@dav-is Oh, last thing: Do you think we need to do anything about unknown attributes, i.e. a provider trying to set attributes that we don't even support? (Although I think that should just be a no-op, i.e. not be written to the database anyway).

**Pull request merged**